### PR TITLE
[Request for Comments] Dvr mode refactor

### PR DESCRIPTION
**Pull request by johnBartos**
- Refactor DVR mode to report duration as Infinity
- Include seekable range as a parameter in time events
- Use seekable range in lieu of duration when in DVR mode
- Call seek with distance from seekableRange when in DVR mode

JW7-3751 (1/2)

https://github.com/jwplayer/jwplayer-commercial/pull/2931 (2/2)

### CODE REVIEW

**Comment by johnBartos** [comment]
test this please

File: src/js/utils/stream-time.js [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+
+], function () {
+    function rewindPosition(amount, currentPosition, seekableRange, streamType) {
+        var pos = 0;
+        var rewindTarget = currentPosition - amount;
+
+        if (streamType === 'DVR') {
+            // When in DVR mode, give the seek position as the distance from the seekable range
+            // The provider will use this to calculate the real time to seek to
+            pos = Math.min(seekableRange - rewindTarget, seekableRange);
```
Comment (egreaves) [inline src/js/utils/stream-time.js]: What happens if `rewindTarget` is larger than `seekableRange`? .ie Someone makes an API call to seek 30s back but you only have 20s of buffer?

File: test/unit/stream-type-util-test.js [hunk 2]
```diff
@@ -3,47 +3,50 @@ define([
...
+        assert.equal(type, 'DVR');
 
-        type = streamTypeUtil.streamType(-120, 0);
-        assert.equal(type, 'DVR', 'streamType with -120 and 0');
+        type = streamTypeUtil.streamType(Infinity, dvrSeekableRange, 0);
+        assert.equal(type, 'DVR');
 
-        type = streamTypeUtil.streamType(-120, undefined);
-        assert.equal(type, 'DVR', 'streamType with 120 and undefined');
+        type = streamTypeUtil.streamType(Infinity, dvrSeekableRange, 0);
```
Comment (egreaves) [inline test/unit/stream-type-util-test.js]: Dupe test

File: src/js/view/view.js [hunk 3]
```diff
@@ -1003,12 +1003,8 @@ define([
             }
         }
 
-        function _setLiveMode(model, duration){
-            var minDvrWindow = model.get('minDvrWindow');
-            var streamType = utils.streamType(duration, minDvrWindow);
+        function _setLiveMode(model, streamType){
```
Comment (egreaves) [inline src/js/view/view.js]: `) {`

File: test/unit/stream-type-util-test.js [hunk 4]
```diff
@@ -3,47 +3,50 @@ define([
...
 
-        type = streamTypeUtil.streamType(-1, minDvrWindow);
-        assert.equal(type, 'LIVE', 'streamType with -1 and 120');
+        type = streamTypeUtil.streamType(Infinity, liveSeekableRange, minDvrWindow);
+        assert.equal(type, 'LIVE');
 
-        type = streamTypeUtil.streamType(Infinity, minDvrWindow);
-        assert.equal(type, 'LIVE', 'streamType with Infinity');
+        type = streamTypeUtil.streamType(Infinity, liveSeekableRange, minDvrWindow);
+        assert.equal(type, 'LIVE');
```
Comment (egreaves) [inline test/unit/stream-type-util-test.js]: Dupe test

**Comment by egreaves** [comment]
My only concern is how the player handles `api.seek` calls for values outside the seekable range. Otherwise, this looks good.

File: src/js/view/components/timeslider.js [hunk 5]
```diff
@@ -152,24 +147,32 @@ define([
...
+                position = (100 - percent) / 100 * seekableRange;
                 this._api.seek(position, reasonInteraction());
             } else {
                 position = percent / 100 * duration;
                 this._api.seek(Math.min(position, duration - 0.25), reasonInteraction());
             }
         },
         showTimeTooltip: function(evt) {
-            var duration = this._model.get('duration');
+            var  duration = this._model.get('duration');
```
Comment (hghazzi) [inline src/js/view/components/timeslider.js]: extra whitespace

File: src/js/controller/model.js [hunk 6]
```diff
@@ -94,11 +94,24 @@ define([
                     this.set('buffer', data.bufferPercent);
                 /* falls through */
                 case events.JWPLAYER_MEDIA_META:
+                    var metaStreamType = utils.streamType(data.duration, data.seekableRange, this.get('minDvrWindow'));
+                    this.set('streamType', metaStreamType);
+                    this.set('seekableRange', data.seekableRange);
```
Comment (robwalch) [inline src/js/controller/model.js]: Do html5 and flash send `seekableRange`? This should require changes to those providers and testing of HLS streams in Safari and in a browser with Flash and primary="flash".

File: src/js/controller/model.js [hunk 7]
```diff
@@ -94,11 +94,24 @@ define([
...
                     var duration = data.duration;
                     if (_.isNumber(duration) && !_.isNaN(duration)) {
                         mediaModel.set('duration', duration);
                         this.set('duration', duration);
                     }
+
+                    // For legacy purposes, report duration as negative video duration when in DVR mode
+                    evt = {
+                        duration: metaStreamType === 'DVR' ? -data.videoDuration : data.duration,
+                        height: data.height,
```
Comment (robwalch) [inline src/js/controller/model.js]: Now all providers need to output `videoDuration` and `seekableRange` for DVR functionality to work.

Should there be a short circuit for legacy providers that return negative duration? Or, documentation for impact to existing providers.

**Comment by robwalch** [comment]
@dannyfinks please include all test links / browsers needed to cover DVR functionality so that @johnBartos and @hghazzi can make sure provider Live/DVR/DOV functionality is improved. In the current state Safari and Flash are impacted by these changes - so would any 3rd party providers with DVR functionality.

**Pull request closed without merging**